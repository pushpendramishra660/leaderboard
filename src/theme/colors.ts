export const colors = {
  white: '#FFFFFF',
  dark: '#666D87BF',
  blue: '#007bff',
  antiFlashWhite: '#F0F0F0',
  venetianRed: '#CC1512',
  grey: '#ccc',
};
