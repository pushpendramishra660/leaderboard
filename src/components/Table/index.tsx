import React from 'react';
import {FlatList} from 'react-native';
import styled, {css} from 'styled-components/native';
import {User} from '../../types/index';
import theme from '../../theme';

interface TableProps {
  data: User[];
  selectedUser?: User | null;
}

const StyledTableContainer = styled.View`
  flex: 1;
  padding: 10px;
`;

const StyledHeader = styled.View`
  flex-direction: row;
  justify-content: space-between;
  padding: 10px 0;
  border-bottom-width: 1px;
  border-bottom-color: ${theme.colors.grey};
`;

const StyledRow = styled.View`
  flex-direction: row;
  justify-content: space-between;
  padding: 10px 0;
  border-bottom-width: 1px;
  border-bottom-color: ${theme.colors.grey};
  width: 100%;
`;

const StyledCell = styled.Text<{isHighlighted?: boolean}>`
  flex: 1;
  text-align: center;
  ${props =>
    props.isHighlighted &&
    css`
      color: ${theme.colors.venetianRed};
    `}
`;

const Table: React.FC<TableProps> = ({data, selectedUser}) => {
  const renderItem = ({item}: {item: User}) => {
    return (
      <StyledRow testID="table-row">
        <StyledCell isHighlighted={item.name === selectedUser?.name}>
          {item.name}
        </StyledCell>
        <StyledCell>{item.rank}</StyledCell>
        <StyledCell>{item.bananas}</StyledCell>
        <StyledCell>{item.subscribed ? 'Yes' : 'No'}</StyledCell>
      </StyledRow>
    );
  };

  return (
    <StyledTableContainer>
      <StyledHeader>
        <StyledCell>Name</StyledCell>
        <StyledCell>Rank</StyledCell>
        <StyledCell>Number of Bananas</StyledCell>
        <StyledCell>Is Searched User</StyledCell>
      </StyledHeader>
      <FlatList
        data={data}
        keyExtractor={item => item.uid}
        renderItem={renderItem}
      />
    </StyledTableContainer>
  );
};

export default Table;
