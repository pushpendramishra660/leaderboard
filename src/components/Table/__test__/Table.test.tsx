import React from 'react';
import renderer from 'react-test-renderer';
import {render} from '@testing-library/react-native';
import Table from '../index';
import {User} from '../../../modals';
import theme from '../../../theme';

describe('<Table />', () => {
  const mockData: User[] = [
    {
      uid: '1',
      name: 'John',
      rank: 1,
      bananas: 5,
      stars: 16,
      subscribed: true,
      lastDayPlayed: '2018-10-22',
      longestStreak: 2,
    },
    {
      uid: '2',
      name: 'Alice',
      rank: 2,
      bananas: 10,
      stars: 16,
      subscribed: false,
      lastDayPlayed: '2019-10-22',
      longestStreak: 2,
    },
  ];

  it('renders Table correctly with provided data', () => {
    const {getByText} = render(<Table data={mockData} />);

    expect(getByText('Name')).toBeTruthy();
    expect(getByText('Rank')).toBeTruthy();
    expect(getByText('Number of Bananas')).toBeTruthy();
    expect(getByText('Is Searched User')).toBeTruthy();
    expect(getByText('John')).toBeTruthy();
  });

  it('correctly highlights the selected user in the table', () => {
    const selectedUser = {
      uid: '1',
      name: 'John',
      rank: 1,
      bananas: 5,
      stars: 16,
      subscribed: true,
      lastDayPlayed: '2018-10-22',
      longestStreak: 2,
    };
    const {getByText} = render(
      <Table data={mockData} selectedUser={selectedUser} />,
    );
    const highlightedCell = getByText('John');
    expect(highlightedCell).toHaveStyle({color: theme.colors.venetianRed});
    expect(highlightedCell.props.style.color).toBe(theme.colors.venetianRed);
  });
  it('displays correct number of items and data in the table', () => {
    const {getByText, getAllByTestId} = render(<Table data={mockData} />);
    expect(getByText('John')).toBeTruthy();
    expect(getByText('Alice')).toBeTruthy();
    const tableRows = getAllByTestId('table-row');
    expect(tableRows.length).toEqual(mockData.length);
  });

  it('Table component matches snapshot', () => {
    const component = renderer.create(<Table data={mockData} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
