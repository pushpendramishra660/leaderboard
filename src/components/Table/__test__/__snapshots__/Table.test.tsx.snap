// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> Table component matches snapshot 1`] = `
<View
  style={
    Object {
      "flexBasis": 0,
      "flexGrow": 1,
      "flexShrink": 1,
      "paddingBottom": 10,
      "paddingLeft": 10,
      "paddingRight": 10,
      "paddingTop": 10,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "#ccc",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 10,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "textAlign": "center",
        }
      }
    >
      Name
    </Text>
    <Text
      style={
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "textAlign": "center",
        }
      }
    >
      Rank
    </Text>
    <Text
      style={
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "textAlign": "center",
        }
      }
    >
      Number of Bananas
    </Text>
    <Text
      style={
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "textAlign": "center",
        }
      }
    >
      Is Searched User
    </Text>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "bananas": 5,
          "lastDayPlayed": "2018-10-22",
          "longestStreak": 2,
          "name": "John",
          "rank": 1,
          "stars": 16,
          "subscribed": true,
          "uid": "1",
        },
        Object {
          "bananas": 10,
          "lastDayPlayed": "2019-10-22",
          "longestStreak": 2,
          "name": "Alice",
          "rank": 2,
          "stars": 16,
          "subscribed": false,
          "uid": "2",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "borderBottomColor": "#ccc",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingBottom": 10,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 10,
              "width": "100%",
            }
          }
          testID="table-row"
        >
          <Text
            isHighlighted={false}
            style={
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "textAlign": "center",
              }
            }
          >
            John
          </Text>
          <Text
            style={
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "textAlign": "center",
              }
            }
          >
            1
          </Text>
          <Text
            style={
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "textAlign": "center",
              }
            }
          >
            5
          </Text>
          <Text
            style={
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "textAlign": "center",
              }
            }
          >
            Yes
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "borderBottomColor": "#ccc",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingBottom": 10,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 10,
              "width": "100%",
            }
          }
          testID="table-row"
        >
          <Text
            isHighlighted={false}
            style={
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "textAlign": "center",
              }
            }
          >
            Alice
          </Text>
          <Text
            style={
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "textAlign": "center",
              }
            }
          >
            2
          </Text>
          <Text
            style={
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "textAlign": "center",
              }
            }
          >
            10
          </Text>
          <Text
            style={
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "textAlign": "center",
              }
            }
          >
            No
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
